pydantic_ai_slim/pydantic_ai/models/grok.py

```diff
@@ -0,0 +1,625 @@
+            return assistant(' '.join(content_parts))
+
+        return None
+
+    def _map_tools(self, model_request_parameters: ModelRequestParameters) -> list[chat_types.chat_pb2.Tool]:
+        """Convert pydantic_ai tool definitions to xAI SDK tools."""
+        tools: list[chat_types.chat_pb2.Tool] = []
+        for tool_def in model_request_parameters.tool_defs.values():
+            xai_tool = tool(
+                name=tool_def.name,
+                description=tool_def.description or '',
+                parameters=tool_def.parameters_json_schema,
+            )
+            tools.append(xai_tool)
+        return tools
+
+    def _get_builtin_tools(self, model_request_parameters: ModelRequestParameters) -> list[chat_types.chat_pb2.Tool]:
+        """Convert pydantic_ai built-in tools to xAI SDK server-side tools."""
+        tools: list[chat_types.chat_pb2.Tool] = []
+        for builtin_tool in model_request_parameters.builtin_tools:
+            if isinstance(builtin_tool, WebSearchTool):
+                tools.append(web_search())
+            elif isinstance(builtin_tool, CodeExecutionTool):
+                tools.append(code_execution())
+            elif isinstance(builtin_tool, MCPServerTool):
+                tools.append(
+                    mcp(
+                        server_url=builtin_tool.url,
+                        server_label=builtin_tool.id,
+                        server_description=builtin_tool.description,
+                        allowed_tool_names=builtin_tool.allowed_tools,
+                        authorization=builtin_tool.authorization_token,
+                        extra_headers=builtin_tool.headers,
+                    )
+                )
+            else:
+                raise UserError(
+                    f'`{builtin_tool.__class__.__name__}` is not supported by `GrokModel`. '
+                    f'Supported built-in tools: WebSearchTool, CodeExecutionTool, MCPServerTool. '
+                    f'If XSearchTool should be supported, please file an issue.'
+                )
+        return tools
+
+    async def request(
+        self,
+        messages: list[ModelMessage],
+        model_settings: ModelSettings | None,
+        model_request_parameters: ModelRequestParameters,
+    ) -> ModelResponse:
+        """Make a request to the Grok model."""
+        # Use injected client or create one in the current async context
+        client = self._client or AsyncClient(api_key=self._api_key)
+
+        # Convert messages to xAI format
+        xai_messages = self._map_messages(messages)
+
+        # Convert tools: combine built-in (server-side) tools and custom (client-side) tools
+        tools: list[chat_types.chat_pb2.Tool] = []
+        if model_request_parameters.builtin_tools:
+            tools.extend(self._get_builtin_tools(model_request_parameters))
+        if model_request_parameters.tool_defs:
+            tools.extend(self._map_tools(model_request_parameters))
```

Comment:
Does the API support `tool_choice`? Look at the logic `OpenAIChatModel` uses to set it; we should do that here as well.